tests/python/unittest/test_tir_transform_lower_cross_thread_reduction.py

```diff
@@ -487,6 +487,149 @@ def lowered_single_reduction_loop_with_block_predicate(
+                )
+
+
+@T.prim_func
+def nested_reduction_loop_with_inner_match_buffers(
+    in0: T.Buffer[(4, 16), "int8"],
+    in1: T.Buffer[(4, 16), "int8"],
+    out: T.Buffer[(4, 4), "int32"],
+) -> None:
+    # body
+    # with T.block("root")
+    for y in T.serial(4):
+        with T.block("C"):
+            yi = T.axis.spatial(4, y)
+            T.reads(in0[yi, 0:16], in1[yi, 0:16])
+            T.writes(out[yi, 0:4])
+            for x in T.serial(4):
+                xr = T.axis.reduce(4, x)
+                with T.init():
+                    for i in T.serial(4):
+                        with T.block("C_init"):
+                            ii = T.axis.spatial(4, i)
+                            T.reads()
+                            T.writes(out[yi, ii])
+                            out[yi, ii] = 0
+                with T.block("C"):
+                    T.reads(
+                        out[yi, xr],
+                        in0[yi, yi * 4 + xr : yi * 4 + xr + 4],
+                        in1[yi, yi * 4 + xr : yi * 4 + xr + 4],
+                    )
+                    T.writes(out[yi, xr])
+                    A = T.match_buffer(
+                        in0[yi, yi * 4 + xr : yi * 4 + xr + 4], [4], dtype="int8", offset_factor=1
+                    )
+                    B = T.match_buffer(
+                        in1[yi, yi * 4 + xr : yi * 4 + xr + 4], [4], dtype="int8", offset_factor=1
+                    )
+                    C = T.match_buffer(out[yi, xr], [1], dtype="int32", offset_factor=1)
+                    A_i8x4: T.int8x4 = A[0:4]
+                    A_i32: T.int32 = T.reinterpret(A_i8x4, dtype="int32")
+                    B_i8x4: T.int8x4 = B[0:4]
+                    B_i32: T.int32 = T.reinterpret(B_i8x4, dtype="int32")
+                    C[0] = A_i32 + B_i32 + C[0]
+
+
+@T.prim_func
+def nested_reduction_loop_with_outer_match_buffers(
+    in0: T.Buffer[(4, 16), "int8"],
+    in1: T.Buffer[(4, 16), "int8"],
+    out: T.Buffer[(4, 4), "int32"],
+) -> None:
+    # body
+    # with T.block("root")
+    for y in T.serial(4):
+        with T.block("C"):
+            yi = T.axis.spatial(4, y)
+            T.reads(in0[yi, 0:16], in1[yi, 0:16])
+            T.writes(out[yi, 0:4])
+            A = T.match_buffer(in0[yi, 0:16], [16], dtype="int8", offset_factor=1)
+            B = T.match_buffer(in1[yi, 0:16], [16], dtype="int8", offset_factor=1)
+            C = T.match_buffer(out[yi, 0:4], [4], dtype="int32", offset_factor=1)
+            for x in T.serial(4):
+                xr = T.axis.reduce(4, x)
+                with T.init():
+                    for i in T.serial(4):
+                        with T.block("C_init"):
+                            ii = T.axis.spatial(4, i)
+                            T.reads()
+                            T.writes(out[yi, ii])
+                            out[yi, ii] = 0
+                with T.block("C"):
```

Comment:
I'm late to the party! Just a quick note: this TVMScript is not valid TIR and I happened to detect it when using the new TVMScript parser which checks more carefully :-)

More specifically, both `T.init()` and `T.axis.reduce` should be placed immediately under a TIR block, while line 611 and line 612 are not :-(

Reply:
If we print out the testcase using `nested_reduction_loop_with_outer_match_buffers.show()`, then the TIR looks like:

```python
# from tvm.script import tir as T
@T.prim_func
def func(in0: T.Buffer[(4, 16), "int8"], in1: T.Buffer[(4, 16), "int8"], out: T.Buffer[(4, 4), "int32"]):
    # body
    # with T.block("root")
    for y in T.serial(4):
        with T.block("C"):
            yi = T.axis.spatial(4, y)
            xr = T.axis.reduce(4, x)
            T.reads(in0[yi, 0 : 16], in1[yi, 0 : 16])
            T.writes(out[yi, 0 : 4])
            A = T.match_buffer(in0[yi, 0 : 16], [16], dtype="int8", offset_factor=1)
            B = T.match_buffer(in1[yi, 0 : 16], [16], dtype="int8", offset_factor=1)
            C = T.match_buffer(out[yi, 0 : 4], [4], dtype="int32", offset_factor=1)
            with T.init():
                for i in T.serial(4):
                    with T.block("C_init"):
                        ii = T.axis.spatial(4, i)
                        T.reads()
                        T.writes(out[yi, ii])
                        out[yi, ii] = 0
            for x in T.serial(4):
                with T.block("C"):
                    T.reads(out[yi, xr], in0[yi, yi * 4 + xr : yi * 4 + xr + 4], in1[yi, yi * 4 + xr : yi * 4 + xr + 4])
                    T.writes(out[yi, xr])
                    A_i8x4: T.int8x4 = A[yi * 4 + xr:yi * 4 + xr + 4]
                    A_i32: T.int32 = T.reinterpret(A_i8x4, dtype="int32")
                    B_i8x4: T.int8x4 = B[yi * 4 + xr:yi * 4 + xr + 4]
                    B_i32: T.int32 = T.reinterpret(B_i8x4, dtype="int32")
                    C[xr] = A_i32 + B_i32 + C[xr]
```

where we may see some use-before-def issues

Reply:
i would love to temporarily exclude this particular TVMScript from testing, but happy to merge it back if you have a follow-up PR to fix :-)

Reply:
Oh interesting! Was not aware of this requirement. I can follow up with a commit in a few days or you can remove the test if its causing issues.

Reply:
https://github.com/apache/tvm/pull/13373/files